##################################################################################################
# Supporting code for Evolution and lineage dynamics of a transmissible cancer in Tasmanian devils
# Author: Kevin Gori
# Date: September 2020
#
# WHAT THIS FILE DOES:
# 1) Generates binned read coverage estimates for samples 1043H and 1044H, which were missing.
# 2) Makes the DOC plot for each sample.
#
# Usage:
# Set working directory to folder containing this script, and these files:
# - devil.fa.fai
# - bins.csv
# - 1043H.interval_depths.txt
# - 1044H.interval_depths.txt
# Then source the file.
#
# Outputs:
# A BED regions file, contig_regions.bed, will be created, and plots
# will be written to 1043H.pdf and 1044H.pdf.
#
# Notes:
# To generate DOC plots for other samples, use the created file contig_regions.bed, and
# run `samtools bedcov -j contig_regions.bed <SAMPLE.BAM> > SAMPLE.interval_depth.txt`
# to create a new interval depths file. This can be used in place of the ones given for
# 1043 and 1044.

require(data.table)

# Set working directory to folder containing this script and files
# - devil.fa.fai
# - 1043H.interval_depths.txt
# - 1044H.interval_depths.txt
#

######################################################
# Generate a mapping between contigs and chromosomes #
# using the reference fasta index                    #
######################################################
filepath = "devil.fa.fai"
fai <- fread(filepath)

colnames(fai) <- c("CONTIG", "LENGTH", "BYTE.OFFSET", "BASES.PER.LINE", "BYTES.PER.LINE")

fn <- function(v) {
    c(0, cumsum(v[1:(length(v)-1)]))
}

fai[, CHROM := substr(CONTIG, 1, 4)]
fai[, OFFSET := fn(LENGTH), by = CHROM]

bed <- fai[(startsWith(CONTIG, "Chr1_") & CONTIG < "Chr1_supercontig_000000399") |
               (startsWith(CONTIG, "Chr2_") & CONTIG < "Chr2_supercontig_000000501") |
               (startsWith(CONTIG, "Chr3_") & CONTIG < "Chr3_supercontig_000000417") |
               (startsWith(CONTIG, "Chr4_") & CONTIG < "Chr4_supercontig_000000317") |
               (startsWith(CONTIG, "Chr5_") & CONTIG < "Chr5_supercontig_000000218") |
               (startsWith(CONTIG, "Chr6_") & CONTIG < "Chr6_supercontig_000000194") |
               (startsWith(CONTIG, "Chrx_") & CONTIG < "Chrx_supercontig_000002377"),
           .(CHROM, START=seq(1, LENGTH, 100000), LENGTH, OFFSET), by=CONTIG]
bed[, END := pmin(START + 100000 - 1, LENGTH)]
bed[, CHROM_START := START + OFFSET]
bed[, CHROM_END := END + OFFSET]
fwrite(bed[, .(CONTIG, START, END)], "contig_regions.bed")

###########################################
# Load the depth estimates:               #
# These were generated by running         #
# samtools bedcov on the bam files, using #
# contig_regions.bed as the regions file  #
###########################################
depths_1043 <- fread("1043H.interval_depths.txt")
depths_1044 <- fread("1044H.interval_depths.txt")

setnames(depths_1043, c("CONTIG", "START", "END", "DEPTH"))
setnames(depths_1044, c("CONTIG", "START", "END", "DEPTH"))

pool_regions <- function(depths_table, bed_table, bins_list) {
    setkey(depths_table, CONTIG, START, END)
    setkeyv(bed_table, key(depths_table))
    x <- depths_table[bed_table, .(Chr = toupper(sub("Chr", "", CHROM)), Start = CHROM_START, End = CHROM_END, DEPTH)]
    x[(End - Start + 1) < 100000, End := End + 1]
    y <- bins_list
    setkey(x, Chr, Start, End)
    setkey(y, Chr, Start, End)

    ol = foverlaps(y, x)
    setkey(ol, Chr, i.Start, i.End)
    ol[, .(DEPTH = sum(DEPTH)), by = key(ol)][, .(Chr, Start = i.Start, End = i.End, DEPTH)]
}

normalise <- function(pooled_table, read_length = 125) {
    # Normalisation -
    #     bedcov depths are sum(coverage per base) per interval.
    #     We want the (estimated) number of reads per 100kb.
    #     1) Divide by length of interval (End - Start + 1), to get avg. cov. per base
    #     2) Divide by read length, to get avg. num. reads per base
    #     3) Divide by two, so depth is per haploid genome (not diploid)
    #     4) Multiply by 100000, to get avg. num. reads per 100kb
    pooled_table[, NORMALISED_DEPTH := as.integer(round(DEPTH / (End - Start + 1) / read_length / 2 * 100000))]
    pooled_table
}


###########################################
# Load the regions as devised by Young Mi #
###########################################
bins <- fread("bins.csv")


##################################
# Normalise the depths, and plot #
##################################
n1043 <- normalise(pool_regions(depths_1043, bed, bins))
n1044 <- normalise(pool_regions(depths_1044, bed, bins))

for (hostname in c("1043H", "1044H")) {
    print(hostname)
    if (hostname == "1043H") {
        data = n1043
    } else {
        data = n1044
    }
    pdf(paste0(hostname,".pdf"), width = 20.04624, height = 11.27601)
    layout_m <- matrix(c(1,1,1,1,1,1,8,
                         2,2,2,2,2,2,2,
                         3,3,3,3,8,8,8,
                         4,4,8,8,8,8,8,
                         5,6,6,7,7,8,8), nrow = 5, byrow = TRUE)
    par(mar = c(5,3,4,0), oma = c(0,0,3,0))
    layout(layout_m, widths = lcm((0. + c(2.51, 1.63, 0.54, 0.53, 0.12, 0.51, 0.15)) /5.99 *50.91745),
           heights = lcm(rep(25.64107 / 5, 5)))

    data[, binCentre := (Start + End) / 2]

    for (chrom in as.character(c(1:6, "X"))) {
        plot(data[Chr == chrom, .(binCentre, NORMALISED_DEPTH)], pch = ".",
             main = paste("Chromosome", chrom), ylab = "Depth", xlab = "Genome Position", ylim = c(0, 1250))
    }
    plot.new()
    mtext(hostname, outer = TRUE)
    dev.off()
}
